child_process: add detached option to .exec()

jessicaquynh opened the pull request:
##### Checklist

- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
Doc, child_process


##### Description of change
<!-- Provide a description of the change below this comment. -->
This PR adds `detached` as an option for `child_process.exec()`
to maintain consistency across the `child_process` functions.

It also adds the `detached` as an option to the documentation 
for `.execSync()`, which was previously undocumented.

Ref: https://github.com/nodejs/node/issues/9146

Fishrock123 commented:
@jessicaquynh seems reasonable to me. 

It will need at least some testing, though.

See our [writing_tests](https://github.com/nodejs/node/blob/master/doc/guides/writing_tests.md) guide if you're not familiar with Node.js core tests. There are probably some tests for `child_process.spawn` in `tests/parallel` that make use of `options.detached` that you could copy from.

cjihrig commented:
Can you add the documentation to `execFile()` as well?

lpinca commented on doc/api/child_process.md:
```diff
@@ -137,6 +137,9 @@ added: v0.1.90
     stdout or stderr - if exceeded child process is killed (Default: `200*1024`)
+  * `detached` {Boolean} Prepare child to run independently of its parent
+    process. Specific behavior depends on the platform, see
+    [`options.detached`][])
```
If I'm not wrong there is a spurious `)` on this line.

lpinca commented on doc/api/child_process.md:
```diff
@@ -625,6 +629,9 @@ added: v0.11.12
     is allowed to run. (Default: `undefined`)
+  * `detached` {Boolean} Prepare child to run independently of its parent
+    process. Specific behavior depends on the platform, see
+    [`options.detached`][])
```
Ditto.

jessicaquynh commented:
I wasn't exactly sure how to write a testing function for `.exec()` to make it hang as noted in the issue without utilizing the `git clone` shell command. Any help would be appreciated.

reecehudson commented:
@jessicaquynh If I read the tests correctly, it would follow after [test-child-process-detached.js)](https://github.com/nodejs/node/blob/master/test/parallel/test-child-process-detached.js) but replace `spawn()` with `execFile()`

reecehudson commented:
I think it would mimic https://github.com/nodejs/node/blob/master/test/parallel/test-child-process-detached.js, but replace `spawn()` with `execFile()`

Fishrock123 commented:
In fact, you may even be able to add it to the same file.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,17 @@
+const assert = require('assert');
+const exec = require('child_process').exec;
+
+const notDetached = exec('git clone https://github.com/private/private', { timeout: 1000, detached: false });
```
There is a good chance that `git` will be on the host system, but I don't think we should rely on `git`, and certainly not GitHub in our tests.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,17 @@
+const notDetached = exec('git clone https://github.com/private/private', { timeout: 1000, detached: false });
+const detached = exec('git clone https://github.com/private2/private2', { detached: true, stdio: ['ignore'] });
+
+notDetached.on('exit', () => {
```
The `'exit'` handlers should be wrapped in `common.mustCall()`. We should probably add assertions for the exit code and signal as well.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,17 @@
+notDetached.on('exit', () => {
+  // check if connected to gitHub
+  if (notDetached._closesGot < 0)
+    assert.equal(notDetached.killed, common.isLinux);
```
Can you replace `assert.equal()` with `assert.strictEqual()`.

jessicaquynh commented:
@cjihrig Sorry I missed your review, I added in the recommendations from @reecehudson and @Fishrock123 to the `test-child-process-detached` and removed the previous test.

cjihrig commented on test/parallel/test-child-process-detached.js:
```diff
@@ -2,24 +2,33 @@
+let execPersistentPid = -1;
 
 const child = spawn(process.execPath, [ childPath ]);
+const execChild = execFile(process.execPath, [ childPath ]);
```
I don't think this actually tests `detached` being passed to `execFile()`. It doesn't happen here, and the fixture calls `spawn()`.

jessicaquynh commented:
So I am actually a bit confused how detached is being tested in `test/parallel/test-child-process-detached.js`. I've done a few experimental tests, flipping `detached` option on and off and I don't notice a difference.

I've read over the documentation, but I don't think that my understanding is completely clear. At the moment, I don't believe that [`.unref()` can be used for `.execFile()`](https://github.com/nodejs/docs/issues/104). The parent will not persist until `child-persistent.js` has completed (which is infinite), so I've put a timeout in the function. But yeah, I think I am just hacking at this code without a full understanding. Any help would be _greatly_ appreciated!!

jasnell commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,26 @@
+  timeout: 1000
+});
+
+execParent.stdout.on('data', function(data) {
```
Nit: perhaps make this an arrow function and wrap it in a `common.mustCall()`

jasnell commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,26 @@
+  assert.notStrictEqual(persistentPid, -1);
+  assert.throws(function(data) {
+    process.kill(execParent.pid);
+  });
```
Should check the specific error here

sam-github commented:
not clear on your question on unref, are you saying your intermediate child process is not exiting before the grand-child?

Also, you seem to be trying to write a test that shows that the detached allows a child to outlive its parent, but that's not what it does on Linux. See the docs. You can do a test where you exec a file with detached, and on non-windows, print its process and session IDs, and the parent can verify that setsid was called in the child. Except that node doesn't have APIs to report the pgid, or sid.

Alternate plan, perhaps you can kill the child by `-child.pid`, proving it has a process group equal to its process ID? see `man 2 setsid` and the process.kill docs

jessicaquynh commented:
@sam-github Sorry for how long it's taken to come back to this. I am beginning to understand a bit more what you mean. When the a child processed is spawned, it inherits the parent process id. When it is set to be detached (at least in Linux), that child is set as the session leader. So when you kill the detached child, the previous session is not affected? At least, that is my pedantic understanding... I appreciate the time you've given!

cjihrig commented on test/fixtures/parent-process-exec-nonpersistent.js:
```diff
@@ -0,0 +1,12 @@
+var execFile = require('child_process').execFile,
```
Please add `'use strict';`

cjihrig replied:
Can you make each `require()` a separate `const` declaration.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,29 @@
+
+const execParent = cp.fork(childPath);
+
+execParent.on('message', common.mustCall((m) => {
```
Can you use a more meaningful variable name, such as `msg` or `message`.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,29 @@
+  persistentPid = m;
+}));
+
+execParent.on('close', common.mustCall((m, sig) => {
```
Similarly, replace `m` with `code`.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,29 @@
+}));
+
+execParent.on('close', common.mustCall((m, sig) => {
+  assert.deepStrictEqual(sig, null);
```
No need to change it, but there is no need for a deep comparison on `sig`.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,29 @@
+const childPath = path.join(common.fixturesDir, '/parent-process-exec-nonpersistent.js');
+
+let persistentPid = -1;
+let execParentCloses = false;
```
This variable isn't necessary. The `common.mustCall()` on the `close` event will handle it.

cjihrig commented on test/parallel/test-child-process-exec-detached.js:
```diff
@@ -0,0 +1,29 @@
+  process.kill(persistentPid);
+}, 100);
+
+process.on('exit', () => {
```
There is no need for this handler.

sam-github commented:
@jessicaquynh 

> When the a child processed is spawned, it inherits the parent process id.

No, it gets its own process ID.

It inherits the parent session and process group.

> When it is set to be detached (at least in Linux), that child is set as the session leader.

And process group leader.

> So when you kill the detached child, the previous session is not affected? 

Killing the detached child never affects the session, wihout detached, it doesn't affect because its not a leader, with detached, it doesn't affect because its not attached to the previous session.

Have you tried this?

> Alternate plan, perhaps you can kill the child by -child.pid, proving it has a process group equal to its process ID? see man 2 setsid and the process.kill docs

Do you know what I mean by using a negative number for kill?

See `man 2 kill`:

> If pid is less than -1, then sig is sent to every process in the process group whose ID is -pid.

Since `setsid()` makes the child a leader of a process group with the same ID as its process ID, my suggestion will test that setsid was called by utilizing one of its effects that is easily observable.

All of above applies *only to unixen*, meaning of detached is completely different on Windows.

jessicaquynh commented:
@sam-github I think I'm beginning to understand what you are getting at a little bit more. I apologize for my redundancy and confusion. I've done a little more studying on *nix processes and have a bit more of a conceptual understanding behind what you're saying!

So essentially, with detached (in *nix), when a child_process is forked, that option ensures that it becomes the session leader.

The session id then becomes that processes's pid, and any subsequent processes spawned from that session inherits that SID. 

Because child_processes can become "orphaned" and continue running even without a parent, if you were to use `.kill()` using the negative pid, it would kill the all processes by process group ID, with absolute value of the negative pid, and thus proving if the child really did become the session leader, and proving detached was called.

(Again, sorry for reiterating, but I just want to make sure I understand clearly).

Additionally, I am wondering exactly what detached does on Windows, and how one can test these behaviours?

sam-github commented:
@jessicaquynh You are well on your way to Unix geekdom, what you said sounds right to me.

As for Windows, I am *not* a Windows dev.... but I believe that in Windows, when a parent process dies, its child is terminated by default. So, on Windows, `.detached` actually allows a child to run longer than its parent.... kindof like the name would indicate. But I'm only mostly sure of that.

It does mean that the Windows test and the non-Windows test will have to be completely different, because the behaviours are completely different.

jessicaquynh commented:
>  You are well on your way to Unix geekdom, what you said sounds right to me.

!! 👍 🎊  :) !!

@sam-github What you've said sounds right to me. I will have to look more into the Windows thing for the test, thanks for all the help so far!

fhinkel commented:
There hasn't been any activity here. I'm closing this. Feel free to reopen (or ping a collaborator) if I closed this in error.

redbaron commented:
It is sad to see this effort lost. What was missing before it can be merged?
